docs/proposals/container-registry-credentials-per-usercluster.md

```diff
@@ -0,0 +1,264 @@
+addons nor user deployed applications by default.
+
+### Proposed feature
+
+Similar to the [SSH key feature][ssh key agent] container registry credentials can be managed in the KKP
+dashboard.
+
+* **KKP admins** can create credentials to be made available for KKP projects
+* **KKP admins** can define default fallback credentials for KKP projects
+* **Project Owners/Editors** can create or select credentials to be made available for user clusters
+* **Project Owners/Editors** can define default fallback credentials for user clusters
+* **User cluster admins** can create or select credentials to be used in all kinds of Pods on the
+user cluster
+
+For details see the [UI section of Implementation](#ui).
+
+There is a possibility to define a default fallback set of credentials in the
+`KubermaticConfiguration` at `spec.userCluster.imagePullSecret`. If there are no credentials
+selected on any level, this will be used as a fallback.
+
+#### Limitations
+
+1. For now only containerd is supported. Because of the [dockershim deprecation][] currently there
+   is no support planned for nodes using docker as container runtime.
+2. Only one credential set per registry can be chosen. Validation will fail if multiple credentials
+   for the same registry are chosen to avoid unexpected behavior.
+
+## Implementation
+
+There are quite some similarities to the implementation of the [SSH key feature][ssh key agent].
+Writing the containerd-config on every worker node of a user cluster based on credentials managed
+and selected in the dashboard sounds quite familiar. This opens up some potential for refactoring
+the current SSH key agent into a more general Agent controlling files on worker node's file system.
+
+### KubermaticConfiguration
+
+A new configuration field to set default fallback registry credentials for any user cluster where no
+credentials are explicitly selected. For consistency this could be in the format of
+`spec.imagePullSecret`:
+
+```yaml
+apiVersion: kubermatic.k8c.io/v1
+kind: KubermaticConfiguration
+metadata:
+  name: <<mykubermatic>>
+  namespace: kubermatic
+spec:
+  userCluster:
+    imagePullSecret: |-
+      {
+        "auths": {
+          "quay.io": {
+            "username": "<<QUAY_USERNAME>>",
+            "password": "<<QUAY_PASSWORD>>",
+          }
+        }
+      }
+```
+
+### Custom Resources
+
+Similar to user ssh keys the `RegistryCredentialSet` is tied to a project, so credentials created in
+one project can be used in all user clusters created in that same project.
+
+`RegistryCredentialSets` wrap `kubernetes.io/dockerconfigjson` typed secrets and have owner
+references to manage visibility and the ability to select credentials.
+
+#### Example
+
+```yaml
+apiVersion: kubermatic.k8c.io/v1
+kind: RegistryCredentialSet
+metadata:
+  name: my-registry-credentials
+  uid: ..
+  ownerReferences:
+  - apiVersion: kubermatic.k8c.io/v1
+    kind: Project
+    name: <<PROJECT_ID>>
+    uid: ..
```

Comment:
Please do not use owner refs for semantical relationships. Owner refs are super clunky and make GitOps much, much harder (thanks to the UID), they cannot be displayed as extra columns in kubectl and are annoying to use in Go code (one has to walk through all the refs to find the owner ref that points to a Project).

A field in the spec would be much better IMHO.